// zox@1.1.2/lib/Responses/EventStreamResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EventStreamResponse {
    constructor(statusCode = 200, headers, onClose) {
        this.headWritten = false;
        this.statusCode = statusCode;
        if (typeof headers === 'object') {
            this.headers = headers;
        }
        if (typeof onClose === 'function') {
            this.onClose = onClose;
        }
    }
    get connected() {
        return !!this.response;
    }
    send(response) {
        response.writeHead(this.statusCode, Object.assign({
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            'Connection': 'keep-alive',
        }, this.headers));
        this.headWritten = true;
        if (this.dataQueue) {
            this.response.write(this.dataQueue);
            this.dataQueue = undefined;
        }
        response.on('close', () => {
            this.response = undefined;
            if (this.onClose) {
                this.onClose();
            }
        });
        this.response = response;
        this.statusCode = undefined;
        this.headers = undefined;
    }
    sendMessage(data, event) {
        data = JSON.stringify(data);
        data = typeof event === 'string' ?
            'event: ' + event + '\n' + 'data: ' + data + '\n\n' :
            'data: ' + data + '\n\n';
        if (this.headWritten) {
            this.response.write(data);
        }
        else {
            this.dataQueue = this.dataQueue ? this.dataQueue + data : data;
        }
    }
    close() {
        this.response.end();
        this.response = undefined;
    }
}
exports.EventStreamResponse = EventStreamResponse;
