// zox@1.1.2/lib/Plugins/PluginManagers/BlockPluginManager.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const ServiceContainer_1 = require("../../ServiceContainer");
const ServicePluginManager_1 = require("../../PluginManagers/ServicePluginManager");
const zox_plugins_1 = require("zox-plugins");
const PluginDiscoveryService_1 = require("../../Services/PluginDiscoveryService");
const serviceKey = Symbol('Block Manager');
const pluginKey = Symbol('Block Manager');
class IBlockPluginManager {
    get serviceKey() {
        return serviceKey;
    }
}
exports.IBlockPluginManager = IBlockPluginManager;
let BlockPluginManager = class BlockPluginManager extends IBlockPluginManager {
    get pluginDefinitions() {
        return this.pluginDiscovery.getPlugins(pluginKey);
    }
    get blockList() {
        const list = [];
        for (const pluginDefinition of this.pluginDefinitions) {
            list.push(pluginDefinition.data.name);
        }
        return list;
    }
    getBlock(name, content, blockData) {
        for (const pluginDefinition of this.pluginDefinitions) {
            if (pluginDefinition.data.name == name) {
                return this.container.create(pluginDefinition.pluginClass, name, content, blockData);
            }
        }
        return null;
    }
};
__decorate([
    ServiceContainer_1.Dependency,
    __metadata("design:type", ServiceContainer_1.IServiceContainer)
], BlockPluginManager.prototype, "container", void 0);
__decorate([
    ServiceContainer_1.Dependency,
    __metadata("design:type", PluginDiscoveryService_1.IPluginDiscoveryService)
], BlockPluginManager.prototype, "pluginDiscovery", void 0);
BlockPluginManager = __decorate([
    ServicePluginManager_1.Service
], BlockPluginManager);
exports.BlockPluginManager = BlockPluginManager;
function Block(options) {
    return zox_plugins_1.PluginSetup(pluginKey, options);
}
exports.Block = Block;
