// zox@1.1.2/lib/Plugins/PluginManagers/RenderEnginePluginManager.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const ServiceContainer_1 = require("../../ServiceContainer");
const ServicePluginManager_1 = require("../../PluginManagers/ServicePluginManager");
const zox_plugins_1 = require("zox-plugins");
const PluginDiscoveryService_1 = require("../../Services/PluginDiscoveryService");
const serviceKey = Symbol('Render Engine Manager');
const pluginKey = Symbol('Render Engine Manager');
class IRenderEnginePluginManager {
    get serviceKey() {
        return serviceKey;
    }
}
exports.IRenderEnginePluginManager = IRenderEnginePluginManager;
let RenderEnginePluginManager = class RenderEnginePluginManager extends IRenderEnginePluginManager {
    get renderEngineInfoList() {
        if (!this._renderEngineInfoList) {
            this._renderEngineInfoList = [];
            const pluginDefinitions = this.pluginDiscovery.getPlugins(pluginKey);
            for (const pluginDefinition of pluginDefinitions) {
                this._renderEngineInfoList.push({
                    renderEngine: this.container.create(pluginDefinition.pluginClass),
                    extensions: pluginDefinition.data,
                });
            }
        }
        return this._renderEngineInfoList;
    }
};
__decorate([
    ServiceContainer_1.Dependency,
    __metadata("design:type", ServiceContainer_1.IServiceContainer)
], RenderEnginePluginManager.prototype, "container", void 0);
__decorate([
    ServiceContainer_1.Dependency,
    __metadata("design:type", PluginDiscoveryService_1.IPluginDiscoveryService)
], RenderEnginePluginManager.prototype, "pluginDiscovery", void 0);
RenderEnginePluginManager = __decorate([
    ServicePluginManager_1.Service
], RenderEnginePluginManager);
exports.RenderEnginePluginManager = RenderEnginePluginManager;
function RenderEngine(...extensions) {
    return zox_plugins_1.PluginSetup(pluginKey, extensions);
}
exports.RenderEngine = RenderEngine;
